src/NHibernate/Type/GenericIdentifierBagType.cs

```diff
@@ -137,10 +137,11 @@ protected override bool AreCollectionElementsEqual(IEnumerable original, IEnumer
 			}
 			else
 			{
+				var originalLookup = iterOriginal.ToLookup(e => e);
 				for (var i = 0; i < targetPc.Count; i++)
 				{
 					var currTarget = targetPc[i];
-					var orgToUse = iterOriginal.First(x => currTarget.Equals(x));
+					var orgToUse = originalLookup[currTarget].First();
```

Comment:
Moreover, this has the advantage of using the default equality comparer for the type, instead of using the base object `Equals`.

I have thought about no taking first but building a map to an original equals elements stack, in order to ensure each original element gets its chance at replacing target element with its own state, since equals original elements may have different states.

Prior to NH-3932, merge would have done that. But well, since bags are un-ordered, which element is taken last is unspecified (original may not be a NHibernate collection, so we cannot tell it will be our internal list ordering). So the resulting state in case equals elements were merge with different states is anyway unpredictable. So I concluded that trying to preserve that behavior was not worth it.